drivers/video/video_stm32_jpeg.c

```diff
@@ -0,0 +1,589 @@
+	k_fifo_put(&data->m2m.in.fifo_out, data->current_in);
+	k_fifo_put(&data->m2m.out.fifo_out, data->current_out);
+
+	/* Try to restart the next processing if needed */
+	ret = stm32_jpeg_start_codec(data->dev);
+	if (ret) {
+		LOG_ERR("Failed to start codec, err: %d", ret);
+		goto out;
+	}
+
+out:
+	k_mutex_unlock(&data->lock);
+}
+
+void HAL_JPEG_ErrorCallback(JPEG_HandleTypeDef *hjpeg)
+{
+	__ASSERT(false, "Got %s", __func__);
+}
+
+/*
+ * This function is called whenever new input data (MCU) must be given in
+ * order to proceed the frame
+ */
+void HAL_JPEG_GetDataCallback(JPEG_HandleTypeDef *hjpeg, uint32_t NbEncodedData)
+{
+	struct stm32_jpeg_data *data =
+			CONTAINER_OF(hjpeg, struct stm32_jpeg_data, hjpeg);
+
+	/* Convert one MCU at a time */
+	stm32_jpeg_nv12_to_ycbcr_mcu(data->current_x_mcu++, data->current_y_mcu,
+				     data->current_in->buffer, data->current_in->buffer +
+				     data->m2m.in.fmt.width * data->m2m.in.fmt.height,
+				     data->mcu_ycbcr, data->m2m.in.fmt.width);
+	if (data->current_x_mcu >= data->m2m.in.fmt.width / MCU_WIDTH) {
+		data->current_x_mcu = 0;
+		data->current_y_mcu++;
+	}
+
+	HAL_JPEG_ConfigInputBuffer(hjpeg, data->mcu_ycbcr, YCBCR_420_MCU_BLOCK_SIZE);
+}
+
+static int stm32_jpeg_get_fmt(const struct device *dev, struct video_format *fmt)
+{
+	struct stm32_jpeg_data *data = dev->data;
+
+	*fmt = fmt->type == VIDEO_BUF_TYPE_INPUT ? data->m2m.in.fmt : data->m2m.out.fmt;
+
+	return 0;
+}
+
+static int stm32_jpeg_set_fmt(const struct device *dev, struct video_format *fmt)
+{
+	struct stm32_jpeg_data *data = dev->data;
+	struct video_common_header *common =
+		fmt->type == VIDEO_BUF_TYPE_INPUT ? &data->m2m.in : &data->m2m.out;
+	struct stm32_jpeg_fmt_conf *conf;
+	int ret = 0;
+
+	/* Validate the settings */
+	conf = stm32_jpeg_get_conf(fmt->pixelformat);
+	if (conf == NULL) {
+		return -EINVAL;
+	}
+	if (fmt->width % conf->hmcu_div || fmt->height % conf->vmcu_div) {
+		LOG_ERR("Format %s: %d pixels width / %d pixels height multiple required",
+			VIDEO_FOURCC_TO_STR(fmt->pixelformat), conf->hmcu_div, conf->vmcu_div);
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	if (common->is_streaming) {
+		ret = -EBUSY;
+		goto out;
+	}
+
+	common->fmt = *fmt;
+
+out:
+	k_mutex_unlock(&data->lock);
+
+	return ret;
+}
+
+static int stm32_jpeg_set_stream(const struct device *dev, bool enable, enum video_buf_type type)
+{
+	struct stm32_jpeg_data *data = dev->data;
+	struct video_common_header *common =
+		type == VIDEO_BUF_TYPE_INPUT ? &data->m2m.in : &data->m2m.out;
+	int ret = 0;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+
+	/* Check that input / output formats are correct */
+	if ((data->m2m.in.fmt.pixelformat == VIDEO_PIX_FMT_JPEG &&
+	     data->m2m.out.fmt.pixelformat == VIDEO_PIX_FMT_JPEG) ||
+	    (data->m2m.in.fmt.pixelformat != VIDEO_PIX_FMT_JPEG &&
+	     data->m2m.out.fmt.pixelformat != VIDEO_PIX_FMT_JPEG)) {
+		LOG_ERR("One of input or output format must be JPEG");
```

Comment:
This is a bit difficult to follow IMO. How about:
```suggestion
        /* shorthands, not *necessary* but helps readability */
        uint32_t in_fmt = data->m2m.in.fmt.pixelformat;
        uint32_t out_fmt = data->m2m.out.fmt.pixelformat;

	if (!((in_fmt == VIDEO_PIX_FMT_JPEG && out_fmt != VIDEO_PIX_FMT_JPEG)
	     || (in_fmt != VIDEO_PIX_FMT_JPEG && out_fmt == VIDEO_PIX_FMT_JPEG))) {
		LOG_ERR("Exactly one of input/output format must be JPEG");
```